tests/test_port.py

```diff
@@ -1,63 +1,69 @@
-import numpy as np
+from numpy.testing import assert_array_equal, assert_array_almost_equal
 import pandas as pd
 
 from sklearn_knn import Raw, Euclidean, GNN
 
+import pytest
 
-def get_check_dist_fn(project, method, ref_or_trg):
-    return f"./tests/data/{method}_{project}_{ref_or_trg}_distances_k5.csv"
 
+class Dataset:
+    def __init__(self, project, method, k=5):
+        self.project = project
+        self.method = method
+        self.k = k
+        
+        self.distances = pd.read_csv(f"./tests/data/{method}_{project}_ref_distances_k{k}.csv")
+        self.neighbors = pd.read_csv(f"./tests/data/{method}_{project}_ref_neighbors_k{k}.csv")
+        self.env_df = pd.read_csv(f"./tests/data/{project}_env.csv")
+        self.spp_df = pd.read_csv(f"./tests/data/{project}_spp.csv")
 
-def get_check_nn_fn(project, method, ref_or_trg):
-    return f"./tests/data/{method}_{project}_{ref_or_trg}_neighbors_k5.csv"
+        cols = [f"K{i+1}" for i in range(k)]
 
+        self.distances = self.distances.loc[:, cols].values
+        self.neighbors = self.neighbors.loc[:, cols].values
+        self.X = self.env_df.iloc[:, 1:].values
+        self.Y = self.spp_df.iloc[:, 1:].values
+        self.ids = self.env_df.iloc[:, 0].values
+        self.spp = self.spp_df.iloc[:, 1:].values
```

Comment:
Oops, good catch! I think sticking with `y` here probably makes more sense for convention and to keep the `Dataset` class more general if we need it for something else. I'll make that change. I'm also open to moving the `Dataset` class (and maybe the dataset fixtures) into another file in the `tests` directory if you have a preference.

Good question on `spp` vs. `species` for the other places where that's used. Also, would there ever be a case where you would want to use something other than a species matrix there and we need to generalize that parameter more (like `y_fit` or `y_modeling` for instance), or is that a fundamental enough part of GNN that we don't need to consider it?

Reply:
I'm good sticking with `y` to match convention.  

I'll answer your `Dataset` class question with another possible test and see if that provides inspiration.  I'd like to test the "new targets" concept, which is essentially a train-test-split and running on the test data.  For testing, I think we can take the top `m` records as training from the existing files and using the `n-m` records for testing.  I can get the corresponding output from yaImpute.  Right now, we haven't done any splitting of the `env` and `spp` files and are testing all plots.  But what do you think about create separate output files like `euc_moscow_ref_neighbors_k5.csv` and `euc_moscow_trg_neighbors_k5.csv` which would have the results from the train and test, respectively?  To the question of the `Dataset` class, would that suggest a different organization or just more attributes to store the `ref` and `trg` files?

I like your idea on renaming the `spp` parameter.  I think we should consider that we have three concepts of "Y" data:

- IDs of the plots themselves
- attributes which will be (optionally) used in setting the transformation during `fit`
- attributes for which will want to predict but are not necessarily used in fitting, although the fit attributes can be part of this list

So if we go with `y`, `y_fit` and `y_predict`, that could make sense to me.  I don't think we need to stick with `spp` or `species` though.  It's a term that is familiar to us, but not necessarily familiar to the broader community.

Reply:
Sorry, just saw in tests that you've kept `ids` and changed `spp` to `y`.  Now I'm conflicted on this one.  In the traditional scikit-learn sense, `y` is meaningful - they are either labels or values of what you are trying to predict.  For us, the ids are only meaningful because they provide the crosswalk to the attributes we are trying to predict.  So maybe now I'm leaning toward explicitly leaving this as `ids` (at least in tests and examples) just to be clear that this is what were outputting with `kneighbors_id`.  I don't want to get high-centered, but I don't want us to be confused every time we look at it as well!

Reply:
Applying a train-test split for our port tests seems like a very good idea to make sure everything's predicting correctly. With that said, I think I would lean towards doing the splits within the test (or possibly using a `pytest` fixture) rather than pre-splitting files. That would mean that the full dataset is easily available for users to run their own splits, which would be more consistent with how `sklearn` handles their datasets. I would also lean towards keeping the neighbor and distance tables internal rather than exposing them in the public dataset, since those are really only needed to test our implementations. That may complicate our testing code a little bit, but it would keep the public-facing datasets a little cleaner and simpler. Does that seem reasonable?

I appreciate the breakdown of the different variations of `y` we're considering. I'm still struggling to keep those separate in my mind... I tend to think things through better with code examples, so am I understanding the use case with the example below?

```python
from sklearnknn import GNN

clf = GNN()

# Fit the ordination between environmental and species, store IDs and structural data for later
clf.fit(X=plot_environmental, y=plot_ids, y_fit=plot_species, y_predict=plot_structural)

# Transform the pixels into CCA space, find nearest neighbors, impute structural data to pixels
pixel_structural = clf.predict(X=pixel_environmental)
```

Or using an example dataset...

```python
from sklearn import GNN
from sklearn.datasets import load_moscow

moscow = load_moscow()
clf = GNN()

clf.fit(X=moscow.data, y=moscow.ids, y_fit=moscow.spp, y_predict=moscow.target)
```

Are those more or less what you had in mind, or am I off base?

Reply:
Yes, I definitely agree that we don't split the environmental and species data and that is handled by the testing code (in a fixture sounds great).  These would be data that is used in testing *and* examples, so it needs to be public-facing.  I also agree that the output neighbors and distances that I'm initially creating in yaImpute are only of use in testing.  

For testing the splits, I started with an 80% train / 20% test, but these were based on the ordered data.  So I output the env/spp data from the training set (first 80% of records) and the env from the testing set (last 20% of records) and saved those for testing.  It looks like we can use `train_test_split` with `shuffle=False` for this (I didn't realize there was a way to NOT randomize the draw).

I love your code examples - thank you for writing those out.  And this is pretty close to what I am thinking.  My one hangup is that `y_predict` is not needed and will not be used in the fit step.  But in terms of getting it into the estimator, I think this makes as much sense as anywhere.  Imagine a scenario like this:
```
clf.fit(X=moscow.data, y=moscow.ids, y_fit=moscow.spp, y_predict=moscow.target)
clf.fit(X=moscow.data, y=moscow.ids, y_fit=moscow.spp, y_predict=new_target_attributes)
```
If you want to use the established model (e.g. neighbors/distances) but want to feed it new attributes to predict, you'd have to call `fit` again even though the model would not change.  Indeed, this is definitely an advantage that we tout with NN methods - if you have a model that's already been fit, you should be able to create new attributes and just come up with new predictions.  Would something like another (non-standard) method to set the target attributes work, like this?
```
clf.fit(X=moscow.data, y=moscow.ids, y_fit=moscow.spp)
clf.set_target_data(y_predict=moscow.target)  # returns `self` like `fit` does
predictions = clf.predict(X=moscow.data)
score = clf.score(X=new_environmental, y=new_target)
```
Thorny issue, this one ...

Reply:
One more thought on what to test.  I think we can just test the 80% train set rather than testing the full set (those would be redundant).  Then we can differentiate those in file naming with `ref` and `trg` labels (like in yaImpute).  So I'm going to replace our current 100% ref files with versions that are from the 80% train set.

Reply:
> (I didn't realize there was a way to NOT randomize the draw)

Me neither, good to know!

> My one hangup is that y_predict is not needed and will not be used in the fit step. But in terms of getting it into the estimator, I think this makes as much sense as anywhere.

This bothered me too, but I came to same conclusion. Your point about changing `y_predict` without refitting is a good one though. I didn't have that use case in mind. I think I like the `set_target_data` idea when a user wants to change the data they're predicting, but I'd be hesitant to *require* that step between fitting and predicting, since most `sklearn` users are familiar with the `fit` then `predict` then `score` workflow and might get tripped up if they try to run `predict` without setting target data.

As an alternative, what do you think about just taking `y_predict` as a parameter of `predict`? No need to refit, no weird unused parameter, no need for a new method. I haven't thought through all the implications, but does this seem reasonable to you?

One thing I'm not clear about that might factor into things is whether we support predicting the `y_fit` if no `y_predict` is given? In other words, if we wanted to fit and predict with structural data, would that be handled implicitly by just passing the structural data to `y_fit`, or will the model fail to predict if it's not explicitly given a `y_predict`? I think I would lean towards keeping it explicit, but if fitting and predicting on the same data is a very common use case, maybe having to specify the same data twice is too much of an inconvenience?

Thorny indeed!

Reply:
This definitely isn't resolved, but I think we've got some good ideas and can continue discussion in #2, so I'm going to call it resolved here